### DISCUSSION TITLE
How to evolve loop bandwidths?

### DISCUSSION THREAD

**Discussion by pfeatherstone** [jgaeddert/liquid-dsp#272] [How do I....?]
If we take [symtrack](https://liquidsdr.org/doc/symtrack/), there are 4 components, each with adaptive loops:
- agc
- symsync (DLL)
- PLL
- equalizer

There is a concept of loop bandwidth, which sets the bandwidth of the LPFs in each of these components.
In the case of symtrack, it initalizes them to:
```
    float agc_bandwidth     = 0.02f  * 0.9f;
    float symsync_bandwidth = 0.001f * 0.9f;
    float eq_bandwidth      = 0.02f  * 0.9f;
    float pll_bandwidth     = 0.001f * 0.9f;
```

Now, how can we dynamically evolve these over time? Is there some metric, possibly using EVM, which allows us to schedule these parameters?
A lot of the literature is unclear on this. Most answers start with "it depends". But it feels like this should be a pretty well understood optimization problem.
Anybody have any ideas?

**Comment by jgaeddert** [comment]
Currently there isn't a way to dynamically evolve these independently. You can, of course, set the overall loop bandwidth of the system with `symtrack_cccf_set_bandwidth()`. The reason I didn't give arbitrary control over the individual loop elements is because they become unstable very quickly.

Two quick points:
  1. You have (rightfully) brought up the idea of using Kalman filtering for this sort of process. I think this is a step in the right direction and would love to explore this further
  2. The components of the `symtrack` object are sample-based (e.g. one sample at a time drives the loops); however it might make sense to move to a block-based approach. This _substantially_ helps with stability issues, but makes the API much more difficult to comprehend. We would need to specify the block size (e.g. 1024 input samples) at object initialization, and then understand that the output won't necessary contain the same number of output symbols.
